packages/itwin/tree-widget/src/components/trees/stateless/models-tree/HierarchyVisibilityHandler.ts

```diff
@@ -0,0 +1,877 @@
+  on: boolean;
+}
+
+interface ChangeModelStateProps {
+  ids: Id64Arg;
+  on: boolean;
+}
+
+/**
+ * Properties for a method of [[IHierarchyBasedVisibilityHandler]] that can be overridden.
+ */
+type OverridableMethodProps<TFunc> = TFunc extends (props: infer TProps) => infer TResult
+  ? TProps & {
+      /** A callback that produces the value from the original implementation. */
+      readonly originalImplementation: () => TResult;
+      /**
+       * Reference to the hierarchy based handler.
+       * @note Calling `getVisibility` or `changeVisibility` of this object invokes the overridden implementation as well.
+       */
+      readonly handler: HierarchyVisibilityHandler;
+    }
+  : never;
+
+/**
+ * Function type for an overridden method of [[IHierarchyBasedVisibilityHandler]].
+ */
+type OverridableMethod<TFunc> = TFunc extends (...args: any[]) => infer TResult ? (props: OverridableMethodProps<TFunc>) => TResult : never;
+
+/**
+ * Functionality of [[IHierarchyBasedVisibilityHandler]] that can be overridden.
+ * Each callback will be provided original implementation and reference to a [[IHierarchyBasedVisibilityHandler]].
+ */
+interface VisibilityHandlerOverrides {
+  getSubjectNodeVisibility?: OverridableMethod<(props: { node: HierarchyNode; ids: Id64Set }) => Promise<VisibilityStatus>>;
+  getModelDisplayStatus?: OverridableMethod<(props: { id: Id64String }) => Promise<VisibilityStatus>>;
+  getCategoryDisplayStatus?: OverridableMethod<(props: GetCategoryStatusProps) => Promise<VisibilityStatus>>;
+  getElementGroupingNodeDisplayStatus?: OverridableMethod<(props: { node: GroupingHierarchyNode }) => Promise<VisibilityStatus>>;
+  getElementDisplayStatus?: OverridableMethod<(props: GetElementStateProps) => Promise<VisibilityStatus>>;
+
+  changeSubjectNodeState?: OverridableMethod<(props: { node: HierarchyNode; ids: Id64Set; on: boolean }) => Promise<void>>;
+  changeModelState?: OverridableMethod<(props: ChangeModelStateProps) => Promise<void>>;
+  changeCategoryState?: OverridableMethod<(props: ChangeCategoryStateProps) => Promise<void>>;
+  changeElementGroupingNodeState?: OverridableMethod<(props: { node: GroupingHierarchyNode; on: boolean }) => Promise<void>>;
+  changeElementState?: OverridableMethod<(props: ChangeElementStateProps) => Promise<void>>;
+}
+
+/**
+ * Properties for [[IHierarchyBasedVisibilityHandler]].
+ * @public
+ */
+export interface HierarchyVisibilityHandlerProps {
+  viewport: Viewport;
+  overrides?: VisibilityHandlerOverrides;
+}
+
+/**
+ * Hierarchy based visibility handler.
+ * When determining visibility for nodes, it should take into account the visibility of their children.
+ * @public
+ */
+export interface HierarchyVisibilityHandler {
+  getVisibilityStatus: (node: HierarchyNode) => Promise<VisibilityStatus>;
+  changeVisibility: (node: HierarchyNode, on: boolean) => Promise<void>;
+  onVisibilityChange: BeEvent<() => void>;
+  // filteredDataProvider?: IFilteredPresentationTreeDataProvider;
```

Comment:
I left in on purpose because this was present in the old visibility handler for determining visibility of subjects. Is it not needed anymore?

Reply:
We won't have an `IFilteredPresentationTreeDataProvider` in the stateless implementation, but we do need to support similar functionality.

Imagine there's a hierarchy like this:
```
+ root subject
+--+ model 1
|  +--+ category 1
|     +--+ element 1
+--+ model 2
|  +--+ category 2
|     +--+ element 2
```

If you filter it to show only `element 1`, you get this:
```
+ root subject
+--+ model 1
|  +--+ category 1
|     +--+ element 1
```

And now, if you click on `root subject` to enable its display, we only want to enable display of what's visible in the hierarchy rather than everything. This means we should enable display of `* 1`, but keep `* 2` as is.

Reply:
If you don't mind, I'd like to do this in a different PR

Reply:
Filed https://github.com/iTwin/viewer-components-react/issues/868